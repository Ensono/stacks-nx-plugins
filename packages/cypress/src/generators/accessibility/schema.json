{
    "$schema": "http://json-schema.org/schema",
    "cli": "nx",
    "$id": "Accessibility",
    "title": "Ensono Stacks cypress Accessibility Generator",
    "description": "The @ensono-stacks/cypress:accessibility generator scaffolds your existing initialised cypress project with example Accessibility tests and config.",
    "type": "object",
    "properties": {
        "project": {
            "type": "string",
            "description": "The name of the project.",
            "alias": "p",
            "$default": {
                "$source": "projectName"
            },
            "x-prompt": "What is the name of the project for this component?"
        }
    },
    "required": ["project"]
}
