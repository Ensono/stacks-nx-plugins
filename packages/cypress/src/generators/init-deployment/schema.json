{
  "$schema": "http://json-schema.org/schema",
  "$id": "InitDeployment",
  "title": "Ensono Stacks Cypress Init Deployment Generator",
  "description": "The @ensono-stacks/cypress:init-deployment generator adds the relevant pipeline and infrastructure code to your workspace.",
  "type": "object",
  "properties": {},
  "required": []
}
