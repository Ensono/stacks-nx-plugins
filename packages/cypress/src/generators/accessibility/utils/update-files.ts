import { tsMorphTree } from '@ensono-stacks/core';
import { joinPathFragments, Tree } from '@nx/devkit';
import {
    MethodDeclaration,
    ObjectLiteralExpression,
    SyntaxKind,
    ImportDeclarationStructure,
    OptionalKind,
} from 'ts-morph';

export const terminalLogAxeBody = `cy.task(
    'log',
    \`\${violations.length} accessibility violation\${
      violations.length === 1 ? '' : 's'
    } \${violations.length === 1 ? 'was' : 'were'} detected\`
  );
  // pluck specific keys to keep the table readable
  const violationData = violations.map(
    ({ id, impact, description, nodes }) => ({
      id,
      impact,
      description,
      nodes: nodes.length,
    })
  );

  cy.task('table', violationData);`;

export function addTerminalLogging(tree: Tree, cypressDirectory: string) {
    const morphTree = tsMorphTree(tree);
    const functionName = 'terminalLogAxe';
    const appNode = morphTree.addSourceFileAtPath(
        joinPathFragments(cypressDirectory, 'src', 'support', 'e2e.ts'),
    );

    if (!appNode.getFunction(functionName)) {
        appNode.addStatements(
            `
import { Result } from 'axe-core';
import addContext from 'mochawesome/addContext';

const cypressGrep = require('@cypress/grep');
cypressGrep();


// Append screenshots on failure to test reports
Cypress.on('test:after:run', (test, runnable) => {
    if (test.state === 'failed') {
        const screenshot = \`../screenshots/\${Cypress.spec.name}/\${runnable.parent?.title} -- \${test.title} (failed).png\`;
        // @ts-ignore
        addContext({ test }, screenshot);
    }
});

export function terminalLogAxe(violations: Result[]) {
    cy.task(
      'log',
      \`\${violations.length} accessibility violation\${violations.length === 1 ? '' : 's'
      } \${violations.length === 1 ? 'was' : 'were'} detected\`
    );
    // pluck specific keys to keep the table readable
    const violationData = violations.map(
      ({ id, impact, description, nodes }) => ({
        id,
        impact,
        description,
        nodes: nodes.length,
      })
    );
  
    cy.task('table', violationData);
  }
  
        `,
        );

        appNode.save();
    }
}

export function updateCypressConfig(tree: Tree, project: string) {
    const sourceFile = tsMorphTree(tree).addSourceFileAtPath(
        joinPathFragments(project, 'cypress.config.ts'),
    );
    const functionName = 'setupNodeEvents';
    const defineConfigExpression = sourceFile
        .getDescendantsOfKind(SyntaxKind.CallExpression)
        .find(callExpression => {
            const expression = callExpression.getExpression();
            return expression && expression.getText() === 'defineConfig';
        })
        ?.getArguments()[0] as ObjectLiteralExpression;
    if (!defineConfigExpression) {
        throw new Error(
            'No defineConfig was found in the application cypress.config.ts file, have you created this using @ensono-stacks/cypress:init?',
        );
    }

    const requiredProperty = defineConfigExpression
        .getPropertyOrThrow('e2e')
        .asKind(SyntaxKind.PropertyAssignment)
        .getInitializerIfKindOrThrow(SyntaxKind.ObjectLiteralExpression);

    const setupNodeEventsMethod: MethodDeclaration =
        requiredProperty
            .getChildrenOfKind(SyntaxKind.MethodDeclaration)
            .find(
                methodDeclaration =>
                    methodDeclaration.getName() === functionName,
            ) ||
        requiredProperty.addMethod({
            name: functionName,
            parameters: [
                {
                    name: 'on',
                },
                {
                    name: 'config',
                },
            ],
        });
    setupNodeEventsMethod.insertStatements(
        setupNodeEventsMethod.getStatements().length,
        `
                on('task', {
                  log(message) {
                    console.log(message);
                    return null;
                  },
                  table(message) {
                    console.table(message);
                    return null;
                  },
                });
            `,
    );
    sourceFile.saveSync();
}
