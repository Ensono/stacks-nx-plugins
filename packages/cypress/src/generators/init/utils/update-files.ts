import { joinPathFragments, Tree, updateJson } from '@nrwl/devkit';

function findOverride(
    json: Array<any>,
    files: Array<string>,
    errorMessage: string,
) {
    return (
        json.find(element =>
            files.every(value => element.files.includes(value)),
        ) ||
        (() => {
            throw new Error(errorMessage);
        })
    );
}

export function updateApplicationLintFile(tree: Tree, path: string) {
    updateJson(tree, joinPathFragments(path, '.eslintrc.json'), eslintjson => {
        const updatedProjectJson = { ...eslintjson };
        const override = findOverride(
            updatedProjectJson.overrides,
            ['*.ts', '*.tsx', '*.js', '*.jsx'],
            'Unable to update the application lint file with with parser options for cypress',
        );
        if (!override.parserOptions) {
            override.parserOptions = {};
        }
        if (!override.parserOptions.project) {
            override.parserOptions.project = [];
        }
        const requiredEntry = joinPathFragments(
            path,
            'cypress',
            'tsconfig(.*)?.json',
        );
        if (!override.parserOptions.project.includes(requiredEntry)) {
            override.parserOptions.project.push(requiredEntry);
        }
        return updatedProjectJson;
    });
}

export function updateTsConfig(tree: Tree, project: string) {
    updateJson(
        tree,
        joinPathFragments(project, 'tsconfig.json'),
        tsConfigJson => {
            const updatedProjectJson = { ...tsConfigJson };
            updatedProjectJson.exclude.push(
                'cypress/**/**',
                'cypress.config.ts',
            );
            updatedProjectJson.references =
                updatedProjectJson.references.filter(
                    reference => reference.path !== './tsconfig.cy.json',
                );
            return updatedProjectJson;
        },
    );
}

export function updateBaseTsConfig(tree: Tree) {
    // Source map set to false due to Cypress hardcoding inlineSourceMap to true
    updateJson(tree, 'tsconfig.base.json', tsConfigJson => {
        const updatedProjectJson = { ...tsConfigJson };
        updatedProjectJson.compilerOptions['sourceMap'] = false;
        return updatedProjectJson;
    });
}
