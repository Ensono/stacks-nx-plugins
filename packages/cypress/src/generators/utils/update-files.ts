import { tsMorphTree } from '@ensono-stacks/core';
import { Tree, updateJson } from '@nrwl/devkit';
import { Project } from 'ts-morph';

let morphTree;

function getMorphTree(tree: Tree) {
    return morphTree ? tsMorphTree(tree) : morphTree;
}

export function updateLintFile(tree: Tree) {
    updateJson(tree, '.eslintrc.json', eslintjson => {
        eslintjson.plugins.push('cypress');
        eslintjson.overrides[1].extends.push('plugin:cypress/recommended');
        return eslintjson;
    });
}
