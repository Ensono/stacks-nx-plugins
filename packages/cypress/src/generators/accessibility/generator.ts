import {
    hasGeneratorExecutedForProject,
    isGeneratorInExecutedListForProject,
    verifyPluginCanBeInstalled,
} from '@ensono-stacks/core';
import {
    addDependenciesToPackageJson,
    formatFiles,
    Tree,
    updateJson,
    joinPathFragments,
} from '@nrwl/devkit';

import { normalizeOptions, addFiles } from '../../utils/test-utils';
import { AXECORE_VERSION, CYPRESSAXE_VERSION } from '../../versions';
import { AccessibilityGeneratorSchema } from './schema';
import { addTerminalLogging, updateCypressConfig } from './utils/update-files';

async function updateDependencies(tree: Tree) {
    return addDependenciesToPackageJson(
        tree,
        {},
        {
            'axe-core': AXECORE_VERSION,
            'cypress-axe': CYPRESSAXE_VERSION,
        },
    );
}

function updateTsConfig(tree: Tree, project: string) {
    updateJson(
        tree,
        joinPathFragments(project, 'tsconfig.json'),
        tsConfigJson => {
            const updatedProjectJson = { ...tsConfigJson };
            if (
                !updatedProjectJson.compilerOptions.types.includes(
                    'cypress-axe',
                )
            ) {
                updatedProjectJson.compilerOptions.types.push('cypress-axe');
            }
            return updatedProjectJson;
        },
    );
}

export default async function accessibilityGenerator(
    tree: Tree,
    options: AccessibilityGeneratorSchema,
) {
    verifyPluginCanBeInstalled(tree, options.project);

    const normalizedOptions = normalizeOptions(tree, options);
    if (
        hasGeneratorExecutedForProject(
            tree,
            options.project,
            'CypressAccessibility',
        )
    )
        return false;
    isGeneratorInExecutedListForProject(
        tree,
        options.project,
        'CypressInit',
        true,
    );

    // generate acessiblity files
    addFiles(
        tree,
        'files',
        __dirname,
        normalizedOptions.projectRoot,
        normalizedOptions,
    );

    // update ts config
    updateTsConfig(tree, normalizedOptions.cypressProject);

    updateCypressConfig(tree, normalizedOptions.projectRoot);

    // add terminal logging funcitonality to e2e.ts
    addTerminalLogging(tree, normalizedOptions.cypressProject);
    await formatFiles(tree);

    return updateDependencies(tree);
}
