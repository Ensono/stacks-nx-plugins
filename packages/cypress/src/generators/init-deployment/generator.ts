import {
    executedDependantGenerator,
    hasGeneratorExecutedForWorkspace,
    verifyPluginCanBeInstalled,
} from '@ensono-stacks/core';
import { formatFiles, Tree } from '@nx/devkit';

import { updateAzureDevopsStages } from './utils/update-azdevops-build';
import { updateTaskctlYaml, updateTasksYaml } from './utils/update-tasks-yamls';

// eslint-disable-next-line consistent-return
export default async function initDeploymentGenerator(tree: Tree) {
    verifyPluginCanBeInstalled(tree);

    if (!executedDependantGenerator(tree, 'WorkspaceDeployment')) return false;
    if (hasGeneratorExecutedForWorkspace(tree, 'CypressInitDeployment'))
        return false;

    updateTaskctlYaml(tree);
    updateTasksYaml(tree);
    updateAzureDevopsStages(tree);

    await formatFiles(tree);
}
