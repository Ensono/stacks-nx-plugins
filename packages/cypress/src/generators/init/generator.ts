import {
    addIgnoreEntry,
    deploymentGeneratorMessage,
    hasGeneratorExecutedForProject,
    tsMorphTree,
} from '@ensono-stacks/core';
import {
    CypressE2EConfigSchema,
    cypressE2EConfigurationGenerator,
} from '@nrwl/cypress/src/generators/cypress-e2e-configuration/cypress-e2e-configuration';
import {
    formatFiles,
    generateFiles,
    getProjects,
    offsetFromRoot,
    Tree,
    readProjectConfiguration,
    addDependenciesToPackageJson,
    updateJson,
} from '@nrwl/devkit';
import { Linter } from '@nrwl/linter';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import path from 'path';

import { updateLintFile } from '../utils/update-files';
import { updateProjectJsonWithHtmlReport } from '../utils/update-targets';
import { CypressGeneratorSchema } from './schema';

interface NormalizedSchema extends CypressGeneratorSchema {
    projectName: string;
    e2eNameForProject: string;
    projectRoot: string;
}

function normalizeOptions(
    tree: Tree,
    options: CypressGeneratorSchema,
): NormalizedSchema {
    const project = getProjects(tree).get(options.project);

    return {
        ...options,
        projectName: project?.name as string,
        e2eNameForProject: `${project?.name as string}-e2e`,
        projectRoot: project?.root as string,
    };
}

function addFiles(
    tree: Tree,
    source: string,
    destination: string,
    options: NormalizedSchema,
) {
    const templateOptions = {
        ...options,
        offsetFromRoot: offsetFromRoot(options.projectRoot),
        template: '',
    };

    generateFiles(
        tree,
        path.join(__dirname, source),
        destination,
        templateOptions,
    );
}

function updateDependencies(tree) {
    return addDependenciesToPackageJson(
        tree,
        {},
        {
            // Needs changing
            '@nrwl/cypress': '15.9.2',
        },
    );
}

export default async function initGenerator(
    tree: Tree,
    options: CypressGeneratorSchema,
) {
    if (hasGeneratorExecutedForProject(tree, options.project, 'CypressInit'))
        return false;
    const normalizedOptions = normalizeOptions(tree, options);
    await cypressE2EConfigurationGenerator(tree, {
        project: normalizedOptions.projectName,
    });
    // update eslint.rc
    updateLintFile(tree);

    // add / remove files
    tree.delete(
        path.join(
            normalizedOptions.e2eNameForProject,
            'src',
            'support',
            'app.po.ts',
        ),
    );
    tree.delete(
        path.join(
            normalizedOptions.e2eNameForProject,
            'src',
            'e2e',
            'app.cy.ts',
        ),
    );
    addFiles(
        tree,
        path.join('files', 'e2e-folder'),
        normalizedOptions.e2eNameForProject,
        normalizedOptions,
    );
    addFiles(
        tree,
        path.join('files', 'root'),
        normalizedOptions.projectRoot,
        normalizedOptions,
    );

    // update targets
    updateProjectJsonWithHtmlReport(
        readProjectConfiguration(tree, normalizedOptions.e2eNameForProject),
        tree,
    );
    // update git ignore
    addIgnoreEntry(tree, '.gitignore', 'Cypress', ['**/test-results']);

    await formatFiles(tree);

    return runTasksInSerial(updateDependencies(tree), () =>
        deploymentGeneratorMessage(
            tree,
            'nx g @ensono-stacks/cypress:init-deployment',
        ),
    );
}
