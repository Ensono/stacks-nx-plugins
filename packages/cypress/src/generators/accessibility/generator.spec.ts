import { tsMorphTree } from '@ensono-stacks/core';
import { createNextApp } from '@ensono-stacks/test';
import { joinPathFragments, readJson, Tree } from '@nx/devkit';
import * as fs from 'fs';
import path from 'path';

import generator from './generator';
import { AccessibilityGeneratorSchema } from './schema';
import { terminalLogAxeBody } from './utils/update-files';
import { checkOneOccurence } from '../../utils/test-utils';
import { AXECORE_VERSION, CYPRESSAXE_VERSION } from '../../versions';
import initGenerator from '../init/generator';

const applicationName = 'application';
const applicationDirectory = `apps/${applicationName}`;
const E2EApplicationDirectory = `apps/${applicationName}-e2e`;

jest.mock('@nx/devkit', () => {
    const actual = jest.requireActual('@nx/devkit');

    return {
        ...actual,
        getProjects: jest.fn(
            () =>
                new Map([
                    [
                        applicationName,
                        {
                            root: applicationDirectory,
                            sourceRoot: applicationDirectory,
                            name: applicationName,
                        },
                    ],
                ]),
        ),
    };
});

function compareToFile(fileInTree, fileToMatchAgainstPath: string) {
    const expectedFileContents = fs
        .readFileSync(path.resolve(__dirname, fileToMatchAgainstPath), 'utf8')
        .replaceAll(/(\r)/gm, '')
        .trim();
    const fileContents = fileInTree.getFullText().trim();
    expect(fileContents).toBe(expectedFileContents);
}

describe('cypress accessibility generator', () => {
    let appTree: Tree;
    let options: AccessibilityGeneratorSchema;
    let project;
    beforeEach(async () => {
        options = {
            project: applicationName,
        };
        appTree = await createNextApp(options.project);
        project = tsMorphTree(appTree);
    });

    it('should raise an error if a cypress app has not yet been added', async () => {
        await generator(appTree, options).catch(error => {
            expect(error.message).toEqual(
                'The dependent CypressInit generator has not been executed',
            );
        });
    });

    it('should raise an error if an invalid project is specified', async () => {
        await generator(appTree, { project: 'nosuchproject' }).catch(error => {
            expect(error.message).toEqual('nosuchproject does not exist.');
        });
    });

    it('should correctly update the setUpNodeEvents if it already exists', async () => {
        await initGenerator(appTree, options);
        const config = project.addSourceFileAtPath(
            joinPathFragments(E2EApplicationDirectory, 'cypress.config.ts'),
        );
        config.replaceWithText(`export default defineConfig({
            ...baseConfig,
            e2e: {
              setupNodeEvents(on, config) {
                on('before:browser:launch', (file) => {
                    //do something
                });
              },
            },
          });`);
        config.saveSync();
        await generator(appTree, options);
        expect(
            appTree
                .read(
                    joinPathFragments(
                        E2EApplicationDirectory,
                        'cypress.config.ts',
                    ),
                )
                .toString(),
        ).toMatchSnapshot();
    });

    describe('should correctly add accessibility', () => {
        beforeEach(async () => {
            await initGenerator(appTree, options);
            await generator(appTree, options);
        });

        it('should update nx.json and tag executed generator true', async () => {
            const nxJson = readJson(appTree, 'nx.json');

            expect(
                nxJson.stacks.executedGenerators.project[
                    options.project
                ].includes('CypressAccessibility'),
            ).toBe(true);
        });

        it('should throw an error if the generator has ran before', async () => {
            const gen = await generator(appTree, {
                ...options,
            });
            expect(gen).toBe(false);
        });

        it('should run successfully and create the accessibility test', async () => {
            // axe-accessibility.spec.ts to be added
            const filePath = joinPathFragments(
                E2EApplicationDirectory,
                'src',
                'e2e',
                'axe-accessibility.cy.ts',
            );
            expect(appTree.exists(filePath)).toBeTruthy();
            compareToFile(
                project.addSourceFileAtPath(filePath),
                './files/src/e2e/axe-accessibility.cy.ts__template__',
            );
        });

        it('should update the package.json with the required dependencies', () => {
            // expect package.json updated
            const packageJson = JSON.parse(
                appTree.read('/package.json', 'utf8'),
            );
            expect(packageJson?.devDependencies).toMatchObject({
                'axe-core': AXECORE_VERSION,
                'cypress-axe': CYPRESSAXE_VERSION,
            });
        });

        it('should update the applications cypress.config.ts', () => {
            expect(
                appTree
                    .read(
                        joinPathFragments(
                            E2EApplicationDirectory,
                            'cypress.config.ts',
                        ),
                    )
                    .toString(),
            ).toMatchSnapshot();
        });

        it('should update the applications e2e.ts support file', () => {
            const filePath = joinPathFragments(
                E2EApplicationDirectory,
                'src',
                'support',
                'e2e.ts',
            );
            const file = project.addSourceFileAtPath(filePath);
            const expectedFunction = file.getFunction('terminalLogAxe');
            const parameters = expectedFunction.getParameters();
            expect(parameters.length).toEqual(1);
            const violationsParameter = expectedFunction
                .getParameters()
                .find(parameter => parameter.getName() === 'violations');
            const parameterType = violationsParameter.getType().getText();
            expect(parameterType).toBe('Result[]');
            expect(expectedFunction?.getBodyText()).toBe(terminalLogAxeBody);
            const importExists = file
                .getImportDeclarations()
                .some(
                    importDecl =>
                        importDecl.getModuleSpecifierValue() === 'axe-core' &&
                        importDecl
                            .getNamedImports()
                            .some(
                                namedImport =>
                                    namedImport.getName() === 'Result',
                            ),
                );
            expect(importExists).toBe(true);
        });

        it('should update the applications cypress tsconfig.json file', () => {
            const tsconfig = readJson(
                appTree,
                joinPathFragments(
                    E2EApplicationDirectory,
                    'src',
                    'tsconfig.json',
                ),
            );
            expect(
                checkOneOccurence(
                    tsconfig.compilerOptions.types,
                    'cypress-axe',
                ),
            ).toBeTruthy();
        });
    });
});
