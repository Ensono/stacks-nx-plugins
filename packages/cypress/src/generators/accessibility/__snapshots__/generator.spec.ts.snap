// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cypress accessibility generator should correctly add accessibility should update the applications cypress.config.ts 1`] = `
"import { nxE2EPreset } from '@nx/cypress/plugins/cypress-preset';

import { defineConfig } from 'cypress';

export default defineConfig({
  e2e: {
    ...nxE2EPreset(__filename, { cypressDir: 'src' }),
    setupNodeEvents(on, config) {
      on('task', {
        log(message) {
          console.log(message);
          return null;
        },
        table(message) {
          console.table(message);
          return null;
        },
      });
    },
  },
});
"
`;

exports[`cypress accessibility generator should correctly update the setUpNodeEvents if it already exists 1`] = `
"export default defineConfig({
  ...baseConfig,
  e2e: {
    setupNodeEvents(on, config) {
      on('before:browser:launch', (file) => {
        //do something
      });

      on('task', {
        log(message) {
          console.log(message);
          return null;
        },
        table(message) {
          console.table(message);
          return null;
        },
      });
    },
  },
});
"
`;
